spx-gui/src/components/asset/preprocessing/split-sprite-sheet/utils.ts

```diff
@@ -0,0 +1,267 @@
+export type Point = [x: number, y: number]
+export type Color = [r: number, g: number, b: number, a: number]
+
+function getPixel(imageData: ImageData, [x, y]: Point): Color {
+  const offset = (imageData.width * y + x) * 4
+  return [
+    imageData.data[offset],
+    imageData.data[offset + 1],
+    imageData.data[offset + 2],
+    imageData.data[offset + 3]
+  ]
+}
+
+function setPixel(imageData: ImageData, [x, y]: Point, color: Color) {
+  const offset = (imageData.width * y + x) * 4
+  imageData.data.set(color, offset)
+}
+
+function equalsColor(a: Color, b: Color, threshold = 10) {
+  return (
+    Math.abs(a[0] - b[0]) < threshold &&
+    Math.abs(a[1] - b[1]) < threshold &&
+    Math.abs(a[2] - b[2]) < threshold &&
+    Math.abs(a[3] - b[3]) < threshold
+  )
+}
+
+export type Region = {
+  startX: number
+  startY: number
+  endX: number
+  endY: number
+}
+
+function hasOverlap(a: Region, b: Region) {
+  return a.startX <= b.endX && a.endX >= b.startX && a.startY <= b.endY && a.endY >= b.startY
+}
+
+function growRegion(imageData: ImageData, satrtPoint: Point, bgColor: Color): Region {
```

Comment:
> 这个是dfs吗

是的

> 会不会造成性能影响？

实现中有一个

```ts
const maxCanvasSize = 1000 * 1000 // to avoid performance issue
```

的限制（最大的计算量是可控的），目前测试看不会有性能问题；另外这个 size 主要是为了控制内存大小，计算基本不是瓶颈

> 可以考虑用现成的图像处理库？

找了一些，没找到合适的；这里的场景（sprite-sheet 切割）比较特殊，更一般的算法（比如 opencv.js 里提供的那些）反而效果不好